src/traits-lowering-to-logic.md

```diff
@@ -0,0 +1,182 @@
+considered true). So the first rule just says "Clone is implemented
+for `usize`". The next rule says "for any type `?T`, Clone is
+implemented for `Vec<?T>` if clone is implemented for `?T`". So
+e.g. if we wanted to prove that `Clone(Vec<Vec<usize>>)`, we would do
+so by applying the rules recursively:
+
+- `Clone(Vec<Vec<usize>>)` is provable if:
+  - `Clone(Vec<usize>)` is provable if:
+    - `Clone(usize)` is provable. (Which is is, so we're all good.)
+    
+But now suppose we tried to prove that `Clone(Vec<Bar>)`. This would
+fail (after all, I didn't give an impl of `Clone` for `Bar`):
+
+- `Clone(Vec<Bar>)` is provable if:
+  - `Clone(Bar)` is provable. (But it is not, as there are no applicable rules.)
+    
+We can easily extend the example above to cover generic traits with
+more than one input type. So imagine the `Eq<T>` trait, which declares
+that `Self` is equatable with a value of type `T`:
+
+```rust
+trait Eq<T> { ... }
+impl Eq<usize> for usize { }
+impl<T: Eq<U>> Eq<Vec<U>> for Vec<T> { }
+```
+
+That could be mapped as follows:
+
+```
+Eq(usize, usize).
+Eq(Vec<?T>, Vec<?U>) :- Eq(?T, ?U).
+```
+
+So far so good. 
+
+## Type-checking normal functions
+
+OK, now that we have defined some logical rules that are able to
+express when traits are implemented and to handle associated types,
+let's turn our focus a bit towards **type-checking**. Type-checking is
+interesting because it is what gives us the goals that we need to
+prove. That is, everything we've seen so far has been about how we
+derive the rules by which we can prove goals from the traits and impls
+in the program; but we are also interesting in how derive the goals
+that we need to prove, and those come from type-checking.
+
+Consider type-checking the function `foo()` here:
+
+```rust
+fn foo() { bar::<usize>() }
+fn bar<U: Eq>() { }
+```
+
+This function is very simple, of course: all it does is to call
+`bar::<usize>()`. Now, looking at the definition of `bar()`, we can see
+that it has one where-clause `U: Eq`. So, that means that `foo()` will
+have to prove that `usize: Eq` in order to show that it can call `bar()`
+with `usize` as the type argument.
+
+If we wanted, we could write a Prolog predicate that defines the
+conditions under which `bar()` can be called. We'll say that those
+conditions are called being "well-formed":
+
+```
+barWellFormed(?U) :- Eq(?U).
+```
+
+Then we can say that `foo()` type-checks if the reference to
+`bar::<usize>` (that is, `bar()` applied to the type `usize`) is
+well-formed:
+
+```
+fooTypeChecks :- barWellFormed(usize).
+```
+
+If we try to prove the goal `fooTypeChecks`, it will succeed:
+
+- `fooTypeChecks` is provable if:
+  - `barWellFormed(usize)`, which is provable if:
+    - `Eq(usize)`, which is provable because of an impl.
+    
+Ok, so far so good. Let's move on to type-checking a more complex function.
+
+## Type-checking generic functions: beyond Horn clauses
```

Comment:
Maybe add an example of implied bounds "in practice" in this section, where we in fact assume `FromEnv(T: Eq)` when type-checking the function below.

Reply:
hmm I think we want to add a new "implied bounds"  section -- beyond what's there now -- that kind of explains in a more "conversational" style. I'll add a placeholder for now, I'm sick of editing.